source/common/network/socket_option_factory.cc

```diff
@@ -33,27 +33,55 @@ SocketOptionFactory::buildTcpKeepaliveOptions(Network::TcpKeepaliveConfig keepal
 std::unique_ptr<Socket::Options> SocketOptionFactory::buildIpFreebindOptions() {
   std::unique_ptr<Socket::Options> options = absl::make_unique<Socket::Options>();
   options->push_back(std::make_shared<Network::AddrFamilyAwareSocketOptionImpl>(
-      Network::Socket::SocketState::PreBind, ENVOY_SOCKET_IP_FREEBIND, ENVOY_SOCKET_IPV6_FREEBIND,
-      1));
+      envoy::api::v2::core::SocketOption::STATE_PREBIND, ENVOY_SOCKET_IP_FREEBIND,
+      ENVOY_SOCKET_IPV6_FREEBIND, 1));
   return options;
 }
 
 std::unique_ptr<Socket::Options> SocketOptionFactory::buildIpTransparentOptions() {
   std::unique_ptr<Socket::Options> options = absl::make_unique<Socket::Options>();
   options->push_back(std::make_shared<Network::AddrFamilyAwareSocketOptionImpl>(
-      Network::Socket::SocketState::PreBind, ENVOY_SOCKET_IP_TRANSPARENT,
+      envoy::api::v2::core::SocketOption::STATE_PREBIND, ENVOY_SOCKET_IP_TRANSPARENT,
       ENVOY_SOCKET_IPV6_TRANSPARENT, 1));
   options->push_back(std::make_shared<Network::AddrFamilyAwareSocketOptionImpl>(
-      Network::Socket::SocketState::PostBind, ENVOY_SOCKET_IP_TRANSPARENT,
+      envoy::api::v2::core::SocketOption::STATE_BOUND, ENVOY_SOCKET_IP_TRANSPARENT,
       ENVOY_SOCKET_IPV6_TRANSPARENT, 1));
   return options;
 }
 
+std::unique_ptr<Socket::Options> SocketOptionFactory::buildLiteralOptions(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::core::SocketOption>& socket_options) {
+  std::unique_ptr<Socket::Options> options = absl::make_unique<Socket::Options>();
```

Comment:
in this pattern you can use 'auto' for the declaration, as it's locally obvious from the template what it must be.

Reply:
Just being consistent with the code around it.

Reply:
What @jmarantz says is best practice for `make_unique` and Envoy.